File: factors.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

/**
 * find_factors - returns a list of factors of a natural number
 * using Fermat's factorization method
 * @n: the natural number to be factorized
 * Return: a pointer to an array of size 2 that contains the factors
 */
long long int *find_factors(long long int n)
{
	long long int *factors = malloc(2 * sizeof(long long int));

	if (n % 2 == 0)
	{
		n = n / 2;
		factors[0] = 2;
	}
	while (n % 2 == 0)
	{
		n = n / 2;
		factors[0] = 2;
	}
	if (n == 1)
	{
		return (factors);
	}
	if (is_prime(n))
	{
		factors[0] = n;
		return (factors);
	}
	long long int a = (long long int)ceil(sqrt(n));
	long long int b = (long long int)floor(sqrt(n));

	while (1)
	{
		long long int c = a * a - b * b;

		if (c == n)
		{
			factors[0] = a - b;
			factors[1] = a + b;
			break;
		}
		else if (c < n)
		{
			a++;
		}
		else
		{
			b--;
		}
	}
	return (factors);
}

/**
 * is_prime - checks if a natural number is prime
 * using the Miller-Rabin primality test
 * @n: the natural number to be checked
 * Return: 1 if the number is prime, 0 otherwise
 */
int is_prime(long long int n)
{
	if (n == 2 || n == 3)
	{
		return (1);
	}
	if (n < 2 || n % 2 == 0)
	{
		return (0);
	}
	long long int r = 0;
	long long int d = n - 1;

	while (d % 2 == 0)
	{
		r++;
		d = d / 2;
	}
	srand(time(NULL));
	long long int a = rand() % (n - 3) + 2;

	/* Compute a^d modulo n */
	long long int x = modular_pow(a, d, n);

	if (x == 1 || x == n - 1)
	{
		return (1);
	}
	for (long long int i = 0; i < r - 1; i++)
	{
		x = modular_pow(x, 2, n);
		if (x == 1)
		{
			return (0);
		}
		if (x == n - 1)
		{
			return (1);
		}
	}
	/* Return 0 */
	return (0);
}

File: factors.py
#!/usr/bin/env python3

import random
import math


"""
A function that returns a list of factors of a natural number
using Fermat's factorization method
"""


def find_factors(n):
    factors = []
    if n % 2 == 0:
        n = n // 2
        factors.append(2)

    while n % 2 == 0:
        n = n // 2
        factors.append(2)
    if n == 1:
        return factors
    if is_prime(n):
        factors.append(n)
        return factors

    a = math.ceil(math.sqrt(n))
    b = math.floor(math.sqrt(n))
    while True:
        c = a * a - b * b
        if c == n:
            factors.append(a - b)
            factors.append(a + b)
            break
        elif c < n:
            a += 1
        else:
            b -= 1
    return factors


"""
A function that checks if a natural number is prime
using the Miller-Rabin primality test
"""


def is_prime(n):
    if n == 2 or n == 3:
        return True
    if n < 2 or n % 2 == 0:
        return False
    r = 0
    d = n - 1
    while d % 2 == 0:
        r += 1
        d = d // 2

    a = random.randint(2, n - 2)
    x = pow(a, d, n)
    if x == 1 or x == n - 1:
        return True
    for _ in range(r - 1):
        x = pow(x, 2, n)
        if x == 1:
            return False
        if x == n - 1:
            return True
    return False


"""
A function that returns a pair of two smaller numbers that multiply to n
"""


def choose_pair(factors):
    p = factors[0]
    q = factors[-1]
    return p, q


"""
A main function that takes a file name as an argument and prints the output
"""


def main(filename):
    try:
        with open(filename, "r") as f:
            lines = f.readlines()
    except FileNotFoundError:
        print("File not found")
        return
    if not lines:
        print("File is empty")
        return
    for line in lines:
        try:
            n = int(line.strip())
        except ValueError:
            print("Invalid input")
            continue
        if n == 0:
            print("0 cannot be factorized")
            continue
        if n < 0:
            print(f"{n} is not a natural number")
            continue
        factors = find_factors(n)
        p, q = choose_pair(factors)
        print(f"{n} = {p} * {q}")


if __name__ == "__main__":
    import sys
    main(sys.argv[1])
